### ISSUE TITLE
[FEATURE] Managing KOReader `lua` files

### ISSUE THREAD

**Issue by shuvashish76** [kyxap/koreader-calibre-plugin#58]
### Feature Summary

Is it possible to backup the lua files

### Motivation

I'm using Syncthing on both PC and mobile. I've created a folder on both pc and phone for ebooks which is sync through Syncthing. On Calibre I use "Connect to folder" feature, copy books there and it gets sync with phone.

The `metadata.epub.lua`, `metadata.pdf.lua` files are stored under respective folders on that ebook folder.

> A Long text column to store the contents of the metadata sidecar as HTML, with Interpret this column as set to HTML.

As per the readme I've used this and it makes the backup but the contents are not exactly same

For example on lua files it's
```
{
    ["annotations"] = {
        [1] = {
            ["chapter"] = "Lesson 1.1: Living a Purpose-Focused Life",
```

But on Calibre I see them as

Normal view
```
{
"annotations": {
"1": {
"chapter": "Lesson 1.1: Living a Purpose-Focused Life",
```
HTML source
```
<div>
<p>{<br>"annotations": {<br>"1": {<br>"chapter": "Lesson 1.1: Living a Purpose-Focused Life",
```

---

### Use Cases

Is it possible to convert them and save them as lua files incase I lose the folder on phone or PC.
Request for the possibility of this feature directly from the plug-in. Also possibility of moving that file to relevant folder for the respective ebooks. If no such folders exist create and move the file. If the file already exist then show _warning_ to the user that it'll re-write the file which will lose existing data.

**Comment by anaxonda** [comment]
Not sure if this is quite what this request is asking, but would be useful to me to have a simple sidecar folder backup feature, from koreader to the calibre folder the book resides in.

**Comment by kyxap** [comment]
Yeah I think it's possible to save sidecar_lua file since currently plugin is doing lua parsing to extract data

**Comment by wfdewith** [comment]
@shuvashish76 doesn't your use case work already? You can currently backup and restore the sidecar files through this plugin when you connect to the synced folder in Calibre. The plugin will take care of converting the JSON back to Lua files when you click "Sync missing to KOReader". The only thing missing is a force sync that will overwrite any existing sidecars.

@anaxonda just so I understand your request correctly.

The current Calibre library looks like this:
```
Library
└── Author Name
    └── Book Title (1)
        ├── Book Title - Author Name.epub
        ├── Book Title - Author Name.pdf
        ├── cover.jpg
        └── metadata.opf
```

and you want the following:
```
Library
└── Author Name
    └── Book Title (1)
        ├── Book Title - Author Name.epub
        ├── Book Title - Author Name.pdf
        ├── Book Title - Author Name.sdr
        │   ├── metadata.epub.lua
        │   └── metadata.pdf.lua
        ├── cover.jpg
        └── metadata.opf
```

Can I ask what the use case is here?

**Comment by harmtemolder** [comment]
@wfdewith I agree with what you're saying in https://github.com/harmtemolder/koreader-calibre-plugin/issues/28#issuecomment-2531791500: it would make sense to store the raw Lua in a separate column, which I believe is what this issue is about. And rewriting the _Sync missing to KOReader_ function to use that raw Lua instead of the JSON

**Comment by anaxonda** [comment]
> Can I ask what the use case is here?

Yes that directory structure is what I was envisioning. Use case would be to be able to open any book from Calibre in Koreader, and have it load any annotations made on an ereader.

I've been finding that using Koreader instead of the built-in Calibre reader much faster.

**Comment by shuvashish76** [comment]
Yes my request was to somehow _automatically_ save those lua files of respective books in Calibre's "Data files"https://manual.calibre-ebook.com/metadata.html#adding-extra-data-files-to-a-book | [comments](https://www.mobileread.com/forums/showthread.php?t=354242)

PS: Somehow missed this notification.

**Comment by shuvashish76** [comment]
> doesn't your use case work already? You can currently backup and restore the sidecar files through this plugin when you connect to the synced folder in Calibre. The plugin will take care of converting the JSON back to Lua files when you click "Sync missing to KOReader".

@wfdewith No #77. Instead of Python parsing Lua and keeping track of whatever way they (KOReader) makes changes in future, I think it's better to keep the exact copy as Calibre's "data files". And the plugin would just manipulate (add/send/remove) that part only.

OR can we offer this as an optional feature?

@kyxap Any thoughts?

**Comment by kyxap** [comment]
yeah I think its fine to save lua file and avoid parsing to json and back, this just need to be done

**Comment by shuvashish76** [comment]
Keeping [bookmarks and highlights ](github.com/kyxap/koreader-calibre-plugin#:~:text=bookmarks%20and%20highlights) as HTML makes sense as you would like to browse/consume such contents but is there any specific reason to keep [contents of the metadata sidecar](https://github.com/kyxap/koreader-calibre-plugin#:~:text=contents%20of%20the%20metadata%20sidecar) as HTML? Can't we simply copy the `.lua` contents directly to Long text column?

Calibre does offer "Interpret this column as" - "**Plain text**" option for "Long text, like comments, not shown in the Tag browser" Column type.

Also IMO "Raw sidecar column" should be mandatory to create for all users.

**Comment by sokolgeo** [comment]
I would agree to have lua file copied as-is / recreated on host Calibre PC. My use case is that I am unable to continue marking up a book on host PC where Calibre installed because .lua file only created on synced devices, not on hots itself!

**Comment by shuvashish76** [comment]
> My use case is that I am unable to continue marking up a book on host PC where Calibre installed because .lua file only created on synced devices, not on hots itself!

You mean changes (e.g. changing book status) to lua file? Only possible via KOReader so I use AppImage for desktop (folder device) . Changes are sync with Syncthing...never used ProgressSync account.

**Comment by sokolgeo** [comment]
I use Syncthing too. Maybe I should go with simple Syncthing-sync of my active reading library across my devices? Then I would probably not need Calibre and this plugin at all?

I guess my largest issue with Syncthing-only setup is my jailbroken Kindle. It runs KoReader and Calibre manages books over USB. But setting up Syncthing might be pain, let alone enabling Wi-Fi on Kindle for Amazon Whisper to start calling home.

**Comment by shuvashish76** [comment]
> Maybe I should go with simple Syncthing-sync of my active reading library across my devices? Then I would probably not need Calibre and this plugin at all?

Yes possible to create separate Calibre library and sync them all but I prefer separate folder-as-device method as I can send desired books without complete Calibre folder bloat (books are stored under _author name - folders only_, each having `cover.jpg`, `metadata.opf `etc). Also I've a huge collection of books so I use Calibre as library management software & e-reader device as reader only not as library i.e. send ~all~ some books only.

> enabling Wi-Fi on Kindle for Amazon Whisper to start calling home.

offtopic: I think if your kindle is jail broken you can add your blocklists to `/etc/hosts` and completely [block Amazon](https://raw.githubusercontent.com/nickspaargaren/no-amazon/master/amazon.txt). Research more this topic, I don't have Kindle so IDK.
